File: day10/day10.py
#!/usr/bin/env python3

OPENED = "([{<"
CLOSED = ")]}>"
POINTS = [3, 57, 1197, 25137]

def check_line(line):
    stack = []
    for c in line:
        idx = OPENED.find(c)
        if idx >= 0:
            stack.append(idx)
        else:
            e = CLOSED[stack.pop()]
            if e != c:
                return stack, POINTS[CLOSED.index(c)]

    return stack, 0

def solve(lines):
    s = 0
    lst = []
    for line in lines:
        stack, points = check_line(line)
        s += points
        if points == 0:
            score = 0
            for i in reversed(stack):
                score = score * 5 + i + 1
            lst.append(score)

    lst = sorted(lst)
    return s, lst[len(lst)//2]

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        part1, part2 = solve(txt.splitlines())
        print("Part1:", part1)
        print("Part2:", part2)

    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

File: day10/day10.cpp
#include <algorithm>
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

static const string OPENED = "([{<";
static const string CLOSED = ")]}>";
static const size_t ERROR_POINTS[] = { 3, 57, 1197, 25137 };

static size_t check_line(const string& line, string& stack)
{
	stack.clear();
	for (auto c : line)
	{
		auto idx = OPENED.find(c);
		if (idx != string::npos)
		{
			stack.push_back(idx);
		}
		else
		{
			if (c != CLOSED[stack.back()])
			{
				return ERROR_POINTS[CLOSED.find(c)];
			}
			stack.pop_back();
		}
	}
	return 0;
}

static pair<size_t, size_t> solve(const vector<string>& lines)
{
	size_t sum = 0;
	vector<size_t> completions;

	string stack;
	for (const auto& line : lines)
	{
		size_t points = check_line(line, stack);
		if (points)
		{
			sum += points;
		}
		else
		{
			size_t score = 0;
			for (auto it = stack.rbegin(); it != stack.rend(); ++it)
			{
				score = score * 5 + *it + 1;
			}
			completions.push_back(score);
		}
	}

	sort(completions.begin(), completions.end());

	return make_pair(sum, completions[completions.size()/2]);
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<string> lines;
	string line;
	while (getline(in, line))
	{
		lines.push_back(move(line));
	}
	in.close();

	auto [part1, part2] = solve(lines);
	fmt::print("Part1: {}\n", part1);
	fmt::print("Part2: {}\n", part2);

	return 0;
}
